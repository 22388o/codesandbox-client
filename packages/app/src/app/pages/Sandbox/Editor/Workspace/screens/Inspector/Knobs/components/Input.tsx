import React from 'react';
import { Input as BaseInput } from '@codesandbox/components';

import css from '@styled-system/css';

export const Input = (props: React.ComponentProps<typeof BaseInput>) => (
  <BaseInput css={css({ borderRadius: '4px', border: 'none' })} {...props} />
);
