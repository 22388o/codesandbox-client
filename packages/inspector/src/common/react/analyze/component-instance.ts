import { SourceLocation } from '@babel/types';
import traverse from '@babel/traverse';
import { CodeRange, FiberSourceInformation } from '../../fibers';
import { parse } from '@babel/parser';

export function analyzeProps(
  code: string,
  range: CodeRange
): FiberSourceInformation {
  function convertBabelPosToPosition(babelPos: SourceLocation): CodeRange {
    return {
      startLineNumber: babelPos.start.line + (range.startLineNumber - 1),
      endLineNumber: babelPos.end.line + (range.startLineNumber - 1),
      startColumnNumber: babelPos.start.column + 1,
      endColumnNumber: babelPos.end.column + 1,
    };
  }

  const lines = code.split('\n');
  const isolatedComponentCode = lines
    .filter((_l, i) => {
      const lineNum = i + 1;
      return lineNum >= range.startLineNumber && lineNum <= range.endLineNumber;
    })
    .join('\n');

  const ast = parse(isolatedComponentCode, {
    plugins: ['jsx', 'typescript'],
    sourceType: 'module',
  });

  let foundComponent = false;
  const foundProps: FiberSourceInformation = { props: {} };
  // @ts-expect-error
  traverse(ast, {
    JSXOpeningElement(path) {
      if (foundComponent) {
        return;
      }

      const location = path.node.loc;
      if (!location) {
        return;
      }

      if (location.start.column < range.startColumnNumber - 1) {
        return;
      }

      foundComponent = true;
      // This means that we're looking at the component we want to look at
      const props = path.get('attributes');

      props.forEach(prop => {
        if (!prop.isJSXAttribute()) {
          return;
        }

        const name = prop.get('name');
        if (!name.isJSXIdentifier()) {
          return;
        }

        const { node: valueNode } = prop.get('value');
        if (valueNode === null || valueNode.loc === null) {
          return;
        }

        const nameNode = name.node;
        const propNode = prop.node;

        if (!propNode.loc || !nameNode.loc) {
          return;
        }

        foundProps.props[nameNode.name] = {
          name: nameNode.name,
          definitionPosition: convertBabelPosToPosition(propNode.loc),
          namePosition: convertBabelPosToPosition(nameNode.loc),
          valuePosition: convertBabelPosToPosition(valueNode.loc),
        };
      });
    },
  });

  return foundProps;
}
