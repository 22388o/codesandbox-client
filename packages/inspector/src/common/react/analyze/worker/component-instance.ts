import ts from 'typescript';
import * as tsutils from 'tsutils';
import { CodeRange, ComponentInstanceData } from '../../../fibers';

export function analyzeComponentInstances(
  sourceFile: ts.SourceFile,
  path: string
): ComponentInstanceData[] {
  function convertTSPosToRange(pos: number, end: number): CodeRange {
    const startRange = ts.getLineAndCharacterOfPosition(sourceFile, pos);
    const endRange = ts.getLineAndCharacterOfPosition(sourceFile, end);
    return {
      startLineNumber: startRange.line + 1,
      endLineNumber: endRange.line + 1,
      startColumnNumber: startRange.character + 1,
      endColumnNumber: endRange.character + 1,
    };
  }

  function getPropInformation(attributes: ts.JsxAttributes) {
    const props: ComponentInstanceData['props'] = {};
    attributes.properties.forEach(property => {
      if (ts.isJsxAttribute(property)) {
        const name = property.name.text;
        props[name] = {
          name: name,
          definitionPosition: convertTSPosToRange(
            property.getStart(),
            property.end
          ),
          namePosition: convertTSPosToRange(
            property.getStart(),
            property.name.end
          ),
          valuePosition: property.initializer
            ? convertTSPosToRange(
                property.initializer.pos,
                property.initializer.end
              )
            : null,
        };
      }
    });
    return props;
  }

  const variableUsage = tsutils.collectVariableUsage(sourceFile);
  function getImportLocation(
    node: ts.Identifier
  ): ComponentInstanceData['importLocation'] | undefined {
    for (let [key, value] of variableUsage) {
      if (!value.uses.some(f => f.location === node)) {
        continue;
      }

      const parent = key.parent;
      if (ts.isImportSpecifier(parent)) {
        const declaration = parent.parent.parent.parent;
        const specifier = declaration.moduleSpecifier;
        if (!ts.isStringLiteral(specifier)) {
          continue;
        }

        const usedName = parent.propertyName || parent.name;
        return {
          importName: usedName.text,
          importPath: specifier.text,
        };
      } else if (ts.isImportClause(parent)) {
        const declaration = parent.parent;
        const specifier = declaration.moduleSpecifier;
        if (!ts.isStringLiteral(specifier)) {
          continue;
        }
        return {
          importName: 'default',
          importPath: specifier.text,
        };
      } else if (
        ts.isVariableDeclaration(parent) &&
        parent.initializer &&
        ts.isIdentifier(parent.initializer)
      ) {
        // Re-export
        return getImportLocation(parent.initializer);
      }
    }

    return undefined;
  }

  const componentInstances: ComponentInstanceData[] = [];

  const visitChild = (node: ts.Node) => {
    if (ts.isJsxOpeningLikeElement(node) && ts.isIdentifier(node.tagName)) {
      const { tagName } = node;
      if (!ts.isIdentifier(tagName) || !tagName.text) {
        return;
      }
      const componentInstance: ComponentInstanceData = {
        name: tagName.text,
        props: getPropInformation(node.attributes),
        importLocation: getImportLocation(tagName),
        location: {
          path,
          codePosition: convertTSPosToRange(node.getStart(), node.end),
        },
      };

      componentInstances.push(componentInstance);

      return;
    }

    node.forEachChild(visitChild);
  };

  visitChild(sourceFile);

  return componentInstances;
}
