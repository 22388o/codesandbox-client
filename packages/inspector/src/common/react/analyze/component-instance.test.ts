import { analyzeProps } from './component-instance';

it('can find props', () => {
  const code = `
  import React from "react";

  import "./styles.css";
  import Input from "./Input";
  import Button from "./Button";
  
  export default function App() {
    return (
      <div className="App">
        <Input />
        <div>
          <Button
            variant={"primary"}
            label="Hello World"
            padding={5}
            onClick={() => {
              console.log("test");
            }}
          />
          <Button />
        </div>
      </div>
    );
  }
  `.trim();

  const result = analyzeProps(code, {
    startLineNumber: 12,
    endLineNumber: 19,
    startColumnNumber: 0,
    endColumnNumber: 0,
  });

  expect(result.props).toMatchSnapshot();
});

it('can find props with children', () => {
  const code = `
  <Button>
    <Input className="test" />
  </Button>
`.trim();

  const result = analyzeProps(code, {
    startLineNumber: 0,
    endLineNumber: 4,
    startColumnNumber: 0,
    endColumnNumber: 0,
  });

  expect(Object.keys(result.props)).toEqual([]);
});

it('can find props from children', () => {
  const code = `<Input><Button className="test" /></Input>`.trim();

  const result = analyzeProps(code, {
    startLineNumber: 1,
    endLineNumber: 1,
    startColumnNumber: 7,
    endColumnNumber: 33,
  });

  expect(Object.keys(result.props)).toEqual(['className']);
});
