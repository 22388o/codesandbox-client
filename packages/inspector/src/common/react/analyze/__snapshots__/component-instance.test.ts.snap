// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can find props 1`] = `
Object {
  "label": Object {
    "definitionPosition": Object {
      "endColumnNumber": 32,
      "endLineNumber": 14,
      "startColumnNumber": 13,
      "startLineNumber": 14,
    },
    "name": "label",
    "namePosition": Object {
      "endColumnNumber": 18,
      "endLineNumber": 14,
      "startColumnNumber": 13,
      "startLineNumber": 14,
    },
    "valuePosition": Object {
      "endColumnNumber": 32,
      "endLineNumber": 14,
      "startColumnNumber": 19,
      "startLineNumber": 14,
    },
  },
  "onClick": Object {
    "definitionPosition": Object {
      "endColumnNumber": 15,
      "endLineNumber": 18,
      "startColumnNumber": 13,
      "startLineNumber": 16,
    },
    "name": "onClick",
    "namePosition": Object {
      "endColumnNumber": 20,
      "endLineNumber": 16,
      "startColumnNumber": 13,
      "startLineNumber": 16,
    },
    "valuePosition": Object {
      "endColumnNumber": 15,
      "endLineNumber": 18,
      "startColumnNumber": 21,
      "startLineNumber": 16,
    },
  },
  "padding": Object {
    "definitionPosition": Object {
      "endColumnNumber": 24,
      "endLineNumber": 15,
      "startColumnNumber": 13,
      "startLineNumber": 15,
    },
    "name": "padding",
    "namePosition": Object {
      "endColumnNumber": 20,
      "endLineNumber": 15,
      "startColumnNumber": 13,
      "startLineNumber": 15,
    },
    "valuePosition": Object {
      "endColumnNumber": 24,
      "endLineNumber": 15,
      "startColumnNumber": 21,
      "startLineNumber": 15,
    },
  },
  "variant": Object {
    "definitionPosition": Object {
      "endColumnNumber": 32,
      "endLineNumber": 13,
      "startColumnNumber": 13,
      "startLineNumber": 13,
    },
    "name": "variant",
    "namePosition": Object {
      "endColumnNumber": 20,
      "endLineNumber": 13,
      "startColumnNumber": 13,
      "startLineNumber": 13,
    },
    "valuePosition": Object {
      "endColumnNumber": 32,
      "endLineNumber": 13,
      "startColumnNumber": 21,
      "startLineNumber": 13,
    },
  },
}
`;
